File: backend/agent_sidecar/schema.js
/**
 * JSON Schema → TypeBox, for the tool schemas `app/tools.py` owns.
 *
 * Pi wants TypeBox for a tool's `parameters`, and Python produces plain JSON
 * Schema. All schemas were dumped (`bench/dump_schemas.py`) and analysed: the
 * converter needs **exactly six keywords** and nothing more —
 *
 *   type · properties · required · items · description · enum
 *
 * so anything else is a signal the schemas changed, and this throws rather than
 * dropping it. A silently dropped constraint means the model sends arguments the
 * tool then rejects, which surfaces as a clarifying question instead of a logged
 * meal — a failure nobody would trace back to a converter.
 *
 * Two cases carry all the risk, and both have their own test:
 *
 * 1. **String enums must use `StringEnum`**, not `Type.Union` /`Type.Literal`.
 *    Pi's `docs/extensions.md` is explicit that the union form breaks Google's
 *    API. Six of them carry an enum, from three distinct definitions —
 *    `discount_split`, `keyword` (shared by reference across four tools), and
 *    `mode`.
 * 2. **One union type exists**: `{"type": ["string", "integer"]}` on `target`, in
 *    `_UPDATE_MEMBER_SCHEMA` and `_DELETE_MEMBER_SCHEMA` (`tools.py:193`, `:216`).
 *    It is the single case a naive converter mistranslates, so it maps to
 *    `Type.Union([...])` deliberately — an *enum* must not, but a genuine type
 *    union has no other spelling.
 */
import { StringEnum, Type } from "@earendil-works/pi-ai";

/** The only keywords the 14 schemas use. Anything else throws. */
export const SUPPORTED_KEYWORDS = new Set([
  "type",
  "properties",
  "required",
  "items",
  "description",
  "enum",
]);

const SCALARS = {
  string: (options) => Type.String(options),
  integer: (options) => Type.Integer(options),
  number: (options) => Type.Number(options),
  boolean: (options) => Type.Boolean(options),
};

function assertSupported(node, path) {
  for (const keyword of Object.keys(node)) {
    if (!SUPPORTED_KEYWORDS.has(keyword)) {
      throw new Error(
        `${path}: unsupported JSON Schema keyword ${JSON.stringify(keyword)}. ` +
          `Add it to schema.js deliberately — dropping it would let the model ` +
          `send arguments the tool rejects.`,
      );
    }
  }
}

/**
 * Convert one JSON Schema node to a TypeBox schema.
 *
 * @param {object} node
 * @param {string} [path] for error messages
 */
export function toTypeBox(node, path = "schema") {
  if (node === null || typeof node !== "object" || Array.isArray(node)) {
    throw new Error(`${path}: expected a JSON Schema object, got ${JSON.stringify(node)}`);
  }
  assertSupported(node, path);

  const options = node.description ? { description: node.description } : {};
  const declared = node.type;

  // The union: {"type": ["string", "integer"]}. Not an enum — a real type union,
  // which is the one thing Type.Union is still the right answer for.
  if (Array.isArray(declared)) {
    const members = declared.map((name) => {
      const build = SCALARS[name];
      if (!build) {
        throw new Error(`${path}: union member ${JSON.stringify(name)} is not a scalar type`);
      }
      return build({});
    });
    return Type.Union(members, options);
  }

  if (declared === "object") {
    const required = new Set(node.required || []);
    const properties = {};
    for (const [key, child] of Object.entries(node.properties || {})) {
      const converted = toTypeBox(child, `${path}.${key}`);
      properties[key] = required.has(key) ? converted : Type.Optional(converted);
    }
    return Type.Object(properties, options);
  }

  if (declared === "array") {
    if (!node.items) {
      throw new Error(`${path}: an array must declare items`);
    }
    return Type.Array(toTypeBox(node.items, `${path}[]`), options);
  }

  if (node.enum) {
    // StringEnum, never Type.Union/Type.Literal: the union form breaks Google's
    // API, and PI_VISION_MODEL is one environment variable away from being a
    // Google model.
    if (declared !== "string") {
      throw new Error(`${path}: enum is only supported on strings, not ${declared}`);
    }
    return StringEnum(node.enum, options);
  }

  const build = SCALARS[declared];
  if (!build) {
    throw new Error(`${path}: unsupported type ${JSON.stringify(declared)}`);
  }
  return build(options);
}

/** Convert a whole `{name: schema}` manifest. */
export function toTypeBoxManifest(schemas) {
  const out = {};
  for (const [name, schema] of Object.entries(schemas)) {
    out[name] = toTypeBox(schema, name);
  }
  return out;
}

File: backend/kernos/data/schema.py
"""The sidecar-safe JSON Schema subset (Phase 5 review F1).

The Pi sidecar converts tool schemas to TypeBox with **exactly six keywords** —
``type, properties, required, items, description, enum`` (``agent_sidecar/schema.js``)
— and throws on anything else while converting the whole manifest, so one
collection schema with ``additionalProperties`` or ``minimum`` would take every tool
of the business down. A collection's schema is therefore checked here, on write,
against the same rules; widening this subset means widening ``schema.js`` and its
tests together.
"""
from __future__ import annotations

import re

SUPPORTED_KEYWORDS = frozenset({"type", "properties", "required", "items", "description", "enum"})
SCALARS = ("string", "integer", "number", "boolean")

SLUG_RE = re.compile(r"^[a-z][a-z0-9_]{0,56}$")          # `{slug}_upsert` stays ≤ 64 chars
DOC_ID_RE = re.compile(r"^[A-Za-z0-9_.:@-]{1,80}$")       # a document id is a URL path segment


class SchemaError(ValueError):
    """A schema outside the sidecar-safe subset; the message names the path and keyword."""


def check_schema(node, path: str = "schema") -> None:
    """Raise :class:`SchemaError` unless ``node`` uses only the safe subset, exactly as
    the sidecar's converter would accept it."""
    if not isinstance(node, dict):
        raise SchemaError(f"{path}: expected a JSON Schema object, got {type(node).__name__}")
    for keyword in node:
        if keyword not in SUPPORTED_KEYWORDS:
            raise SchemaError(f"{path}: unsupported JSON Schema keyword {keyword!r} "
                              f"(the sidecar converts only {sorted(SUPPORTED_KEYWORDS)})")
    declared = node.get("type")
    if isinstance(declared, list):
        for name in declared:
            if name not in SCALARS:
                raise SchemaError(f"{path}: union member {name!r} is not a scalar type")
        if "enum" in node:
            raise SchemaError(f"{path}: enum is only supported on strings")
        return
    if declared == "object":
        props = node.get("properties")
        if not isinstance(props, dict):
            raise SchemaError(f"{path}: an object must declare properties")
        for req in node.get("required") or []:
            if req not in props:
                raise SchemaError(f"{path}: required names unknown property {req!r}")
        for key, child in props.items():
            check_schema(child, f"{path}.{key}")
        return
    if declared == "array":
        if "items" not in node:
            raise SchemaError(f"{path}: an array must declare items")
        check_schema(node["items"], f"{path}[]")
        return
    if "enum" in node:
        if declared != "string":
            raise SchemaError(f"{path}: enum is only supported on strings, not {declared!r}")
        if not isinstance(node["enum"], list) or not all(isinstance(v, str) for v in node["enum"]):
            raise SchemaError(f"{path}: enum values must be strings")
        return
    if declared not in SCALARS:
        raise SchemaError(f"{path}: unsupported type {declared!r}")


def check_collection_schema(schema, *, key: str, indexed: list[str]) -> None:
    """A collection's schema: an object in the safe subset whose ``key`` is a required
    string property and whose ``indexed`` fields are properties."""
    check_schema(schema)
    if schema.get("type") != "object":
        raise SchemaError("schema: a collection schema must be an object")
    props = schema["properties"]
    if key not in props:
        raise SchemaError(f"key {key!r} is not a property of the schema")
    if props[key].get("type") != "string":
        raise SchemaError(f"key {key!r} must be a string property")
    if key not in (schema.get("required") or []):
        raise SchemaError(f"key {key!r} must be required")
    for field in indexed:
        if field not in props:
            raise SchemaError(f"indexed field {field!r} is not a property of the schema")
